### PR TITLE
[SteamOverHolland] fixes revenue bugs

### LINKED ISSUE #11150: Steam Over Holland - Unable to enter game
**What's your Game ID?** https://18xx.games/game/174379

**How to Reproduce the Bug?**

**Which Browser / Platform?** desktop and mobile.  Android and Win 11 MSEdge

### LINKED ISSUE #11155: [Steam over Holland] unable to run to Vlissengen after acquiring p4
https://18xx.games/game/174357

P4 is making this route illegal for some reason, I can run to Vlissengen without buying in the private just fine???

### LINKED ISSUE #11157: [Steam Over Holland] Unable to enter game 174384
Game 174384 will not load. Last action unable to be retrieved; screen stuck on “game loading” screen.

**What's your Game ID?**

(https://18xx.games/game/174384)

-->

**How to Reproduce the Bug?**

uncertain: game stopped loading on another person’s turn.

edit: wrong game number

### LINKED ISSUE #11164: [Steam over Holland] company can not run a route over the ferry while owing p3
https://18xx.games/game/174408

This is a similar bug to the one with being unable to run to Vlissengen while owning p4

### PR DESCRIPTION
**Pull request by philcampeau**
Fixes #11164
Fixes #11155
Fixes #11150 
Fixes #11157 

## Before clicking "Create"

- [x] Branch is derived from the latest `master`
- [x] Add the `pins` or `archive_alpha_games` label if this change will break existing games
- [x] Code passes linter with `docker compose exec rack rubocop -a`
- [x] Tests pass cleanly with `docker compose exec rack rake`

### Explanation of Change

Code I had for checking route bonuses doesn't seem to work. I replaced it with essentially the exact code from 1867 applying the same bonuses (simplified due to not needing the hexes to be unique).

Archive shouldn't be needed, since these routes would have simply been blocked before.

### CODE REVIEW

File: lib/engine/game/g_steam_over_holland/game.rb [hunk 1]
```diff
@@ -369,12 +369,13 @@ def check_distance(route, visits, train = nil)
         def revenue_for(route, stops)
           revenue = super
 
-          abilities = Array(abilities(route.corporation, :hex_bonus))
-          return revenue if abilities.empty?
+          route.corporation.companies.each do |company|
+            abilities(company, :hex_bonus) do |ability|
```
Comment (ollybh) [inline lib/engine/game/g_steam_over_holland/game.rb:373]: You don't need nested iterators here. Passing a Corporation object to `Game.abilities` will return matching abilities from both the corporation and any companies that it owns. See `Corporation.all_abilities`: https://github.com/tobymao/18xx/blob/38a729965ef394d5cd3ace16e45952672fa64aac/lib/engine/corporation.rb#L252-L262

If you do have the nested iterator, you need to remember that `abilities` can return nil, a single ability, or an array of abilities. The inner iterator will fail unless the company has two or more hex_bonus abilities.

File: lib/engine/game/g_steam_over_holland/game.rb [hunk 2]
```diff
@@ -369,12 +369,13 @@ def check_distance(route, visits, train = nil)
...
+              revenue += stops.sum { |s| ability.hexes.include?(s.hex.id) ? ability.amount : 0 }
```
Comment (ollybh) [inline lib/engine/game/g_steam_over_holland/game.rb:374]:
```suggestion
              revenue += ability.amount * stops.count { |s| ability.hexes.include?(s.hex.id) }
```

**Review by ollybh — CHANGES_REQUESTED** [review]

File: lib/engine/game/g_steam_over_holland/game.rb (hunk 2 above)
Comment (philcampeau) [reply lib/engine/game/g_steam_over_holland/game.rb:374]: done

File: lib/engine/game/g_steam_over_holland/game.rb [hunk 3]
```diff
@@ -369,12 +369,11 @@ def check_distance(route, visits, train = nil)
         def revenue_for(route, stops)
           revenue = super
 
-          abilities = Array(abilities(route.corporation, :hex_bonus))
-          return revenue if abilities.empty?
+          abilities(route.corporation, :hex_bonus) do |ability|
```
Comment (ollybh) [inline lib/engine/game/g_steam_over_holland/game.rb:372]:
```suggestion
          Array(abilities(route.corporation, :hex_bonus)).each do |ability|
```

File: lib/engine/game/g_steam_over_holland/game.rb (hunk 1 above)
Comment (philcampeau) [reply lib/engine/game/g_steam_over_holland/game.rb]: done

File: lib/engine/game/g_steam_over_holland/game.rb (hunk 3 above)
Comment (philcampeau) [reply lib/engine/game/g_steam_over_holland/game.rb]: done

**Pull request merged**